sdk/program/src/bignum.rs

```diff
@@ -0,0 +1,851 @@
+                }
+            };
+            unsafe {
+                sol_bignum_mod_inv(
+                    &lhs_ffi_in as *const _ as *const u64,
+                    &rhs_ffi_in as *const _ as *const u64,
+                    &mut bn_syscall_result as *mut _ as *mut u64,
+                );
+                value.set_len(bn_syscall_result.data_len as usize);
+                Self {
+                    is_negative: bn_syscall_result.is_negative,
+                    value,
+                }
+            }
+        }
+    }
+
+    /// Log a `BigNum` from a program to the solana logs
+    pub fn log(&self) {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            crate::program_stubs::sol_log(&self.value.to_dec_str().unwrap());
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            extern "C" {
+                fn sol_bignum_log(lhs_ffi_in_addr: *const u64) -> u64;
+            }
+            let lhs_ffi_in = {
+                FfiBigNumber {
+                    data: self.value.as_ptr() as *const _ as u64,
+                    data_len: self.value.len() as u64,
+                    is_negative: self.is_negative,
+                }
+            };
+            unsafe { sol_bignum_log(&lhs_ffi_in as *const _ as *const u64) };
+        }
+    }
+}
+
+impl PartialEq for BigNumber {
+    fn eq(&self, other: &Self) -> bool {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            self.value == other.value && self.value.is_negative() == other.value.is_negative()
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            self.value == other.value && self.is_negative == other.is_negative
+        }
+    }
+}
+impl Clone for BigNumber {
+    fn clone(&self) -> Self {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            match openssl::bn::BigNum::from_slice(&self.value.to_vec()) {
+                Ok(bn) => Self { value: bn },
+                Err(_) => panic!(),
+            }
+        }
+        #[cfg(target_arch = "bpf")]
+        Self {
+            value: self.value.clone(),
+            is_negative: self.is_negative,
+        }
+    }
+}
+
+impl Default for BigNumber {
+    fn default() -> Self {
+        BigNumber::new()
+    }
+}
+impl fmt::Debug for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            write!(f, "{}", self.value.to_dec_str()?)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+impl fmt::Display for BigNumber {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        #[cfg(not(target_arch = "bpf"))]
+        {
+            // let braw: &BigNum = unsafe { &*(self.0 as *mut BigNum) };
+            write!(f, "{}", self.value.to_dec_str()?)
+        }
+        #[cfg(target_arch = "bpf")]
+        {
+            write!(f, "{}", 0)
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
Be nice if the tests covered all the methods

Reply:
@jackmay So our thoughts were that the unit tests in the source file are non-BPF so we'd just be testing OpenSSL again vs. our BPF equivalents ... thoughts?

Reply:
Yeah, sounds good